tests/unit_tests/tests_launch/test_launch_kubernetes.py

```diff
@@ -0,0 +1,164 @@
+    __setattr__ = dict.__setitem__
+    __delattr__ = dict.__delitem__
+
+
+class MockPodList:
+    def __init__(self, pods):
+        self.pods = pods
+
+    @property
+    def items(self):
+        return self.pods
+
+
+class MockBatchV1Api:
+    def __init__(self, mock_api_client, jobs):
+        self.jobs = jobs
+
+    def read_namespaced_job(self, name, namespace):
+        return self.jobs[name]
+
+
+class MockCoreV1Api:
+    def __init__(self, mock_api_client, pods):
+        # self.context = mock_api_client["context_name"]
+        self.pods = pods
+
+    def list_namespaced_pod(self, label_selector, namespace):
+        ret = []
+        k, v = label_selector.split("=")
+        if k == "job-name":
+            for pod in self.pods.items:
+                if pod.job_name == v:
+                    ret.append(pod)
+        return MockPodList(ret)
+
+
+def pods(job_name):
+    return MockPodList(
+        [
+            MockDict(
+                {
+                    "metadata": MockDict(
+                        {
+                            "name": "pod1",
+                        }
+                    ),
+                    "job_name": job_name,
+                    "log": "test log string",
+                }
+            )
+        ]
+    )
+
+
+def setup_mock_kubernetes_client(monkeypatch, jobs, pods, mock_job_base):
+    def mock_create_from_yaml(yaml_objects, jobs_dict, mock_status):
+        jobd = yaml_objects[0]
+        name = jobd["metadata"]["name"]
+        if not name:
+            name = jobd["metadata"]["generateName"] + "testname"
+
+        metadata = MockDict(jobd["metadata"])
+        metadata.labels = metadata.get("labels", {})
+        metadata.labels["job-name"] = name  # assign name
+        job_spec = MockDict(jobd["spec"])
+        job_spec.backoff_limit = job_spec.get("backoffLimit", 6)  # kube defaults
+        job_spec.completions = job_spec.get("completions", 1)
+        job_spec.parallelism = job_spec.get("parallelism", 1)
+        job_spec.suspend = job_spec.get("suspend", False)
+        pod_spec = MockDict(jobd["spec"]["template"]["spec"])
+        pod_spec.restart_policy = pod_spec.get("restartPolicy", "Never")
+        pod_spec.preemption_policy = pod_spec.get(
+            "preemptionPolicy", "PreemptLowerPriority"
+        )
+        pod_spec.node_name = pod_spec.get("nodeName", None)
+        pod_spec.node_selector = pod_spec.get("nodeSelector", {})
+        pod_spec.containers = pod_spec.get("containers")
+        for i, cont in enumerate(pod_spec.containers):
+            pod_spec.containers[i] = MockDict(cont)
+
+        job_spec.template = MockDict(
+            {
+                "spec": pod_spec,
+            }
+        )
+        mock_job = MockDict(
+            {
+                "status": mock_status,
+                "spec": job_spec,
+                "metadata": MockDict(metadata),
+            }
+        )
+        jobs_dict[name] = mock_job
+        return [[mock_job]]
```

Comment:
im curious about style here, but I sometimes like to put internal functions like this after the top level function code. python doesn't care about order and you can more easily tell when that function stops. total style thing and im pretty sure we do both in the codebase.

Reply:
As in you would do the monkey path.setattr stuff at the top of the function, then have the `mock_create_from_yaml` definition after...